src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -4996,9 +4996,20 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           typedSelect(tree, typedSelectOrSuperQualifier(qual), nme.CONSTRUCTOR)
         case Select(qual, name) =>
           if (Statistics.canEnable) Statistics.incCounter(typedSelectCount)
-          val qualTyped = checkDead(typedQualifier(qual, mode))
+          def typeQual = checkDead(typedQualifier(qual, mode))
+          val (qualTyped, isJavaTypeSelection) = {
+            if (name.isTermName) (typeQual, false)
+            else {
+              qual match {
+                case id: Ident if context.unit.isJava =>
```

Comment:
A comment would be good here, e.g.

```
scala/bug#10490: given a Scala class A { class B }, in mixed compilation, a reference A.B in a Java source
would cause a "not found: value A". The `JavaTypeSelection` mode makes `typedIdent` type check A as
a type if no value is found. In case A has a companion, it works thanks to the fix of scala/bug#3120 in
typedSelectInternal.
```

Reply:
I'll add it 👍